sw/device/lib/dif/dif_i2c.c

```diff
@@ -0,0 +1,681 @@
+                                             bool sda) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t override_val =
+      mmio_region_read32(i2c->base_addr, I2C_OVRD_REG_OFFSET);
+  override_val = bitfield_set_field32(
+      override_val, (bitfield_field32_t){
+                        .mask = 1, .index = I2C_OVRD_SCLVAL, .value = scl,
+                    });
+  override_val = bitfield_set_field32(
+      override_val, (bitfield_field32_t){
+                        .mask = 1, .index = I2C_OVRD_SDAVAL, .value = sda,
+                    });
+  mmio_region_write32(i2c->base_addr, I2C_OVRD_REG_OFFSET, override_val);
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_override_sample_pins(const dif_i2c_t *i2c,
+                                              uint16_t *scl_samples,
+                                              uint16_t *sda_samples) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t samples = mmio_region_read32(i2c->base_addr, I2C_VAL_REG_OFFSET);
+  samples >>= I2C_VAL_SCL_RX_OFFSET;
+  if (scl_samples != NULL) {
+    uint32_t val = samples >> I2C_VAL_SCL_RX_OFFSET;
+    *scl_samples = val & I2C_VAL_SCL_RX_MASK;
+  }
+
+  if (sda_samples != NULL) {
+    uint32_t val = samples >> I2C_VAL_SDA_RX_OFFSET;
+    *sda_samples = val & I2C_VAL_SDA_RX_MASK;
+  }
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_get_fifo_levels(const dif_i2c_t *i2c,
+                                         uint8_t *fmt_fifo_level,
+                                         uint8_t *rx_fifo_level) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t values =
+      mmio_region_read32(i2c->base_addr, I2C_FIFO_STATUS_REG_OFFSET);
+  if (fmt_fifo_level != NULL) {
+    uint32_t val = values >> I2C_FIFO_STATUS_FMTLVL_OFFSET;
+    *fmt_fifo_level = val & I2C_FIFO_STATUS_FMTLVL_MASK;
+  }
+  if (rx_fifo_level != NULL) {
+    uint32_t val = values >> I2C_FIFO_STATUS_RXLVL_OFFSET;
+    *rx_fifo_level = val & I2C_FIFO_STATUS_RXLVL_MASK;
+  }
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_read_byte(const dif_i2c_t *i2c, uint8_t *byte) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t rdata =
+      mmio_region_read_mask32(i2c->base_addr, I2C_RDATA_REG_OFFSET,
+                              I2C_RDATA_RDATA_MASK, I2C_RDATA_RDATA_OFFSET);
+  if (byte != NULL) {
+    *byte = rdata;
+  }
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_write_byte_raw(const dif_i2c_t *i2c, uint8_t byte,
+                                        dif_i2c_fmt_flags_t flags) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+  // Validate that "write only" flags and "read only" flags are not set
+  // simulataneously. `start` is valid for both modes.
+  bool has_write_flags = flags.stop || flags.nak_ok;
+  bool has_read_flags = flags.read || flags.read_cont;
+  if (has_write_flags && has_read_flags) {
```

Comment:
Good observations, though I think your first one is incorrect: `READ` implies `STOP`, unless `RCONT` is set, AIUI.

I'll update the logic to reflect this though.

Reply:
`READ` should not imply `STOP` because the host should be able to send a `START` (repeated start) to keep communicating with the target device without releasing the bus.

Reply:
Just wanted to make sure that you saw my previous comment on `READ` and `STOP` (please also see the state machine in the tech spec).